catalog/core/catalog-core-api-impl/src/test/java/ddf/catalog/util/impl/QueryResultPaginatorTest.java

```diff
@@ -0,0 +1,395 @@
+        assertThat(paginator.hasNext(), equalTo(false));
+        assertThat(paginator.getBufferSize(), equalTo(0));
+    }
+
+    @Test
+    public void testIndexingAcrossPages() {
+        int pageSize = 5;
+        setStartIndexAndPageSize(10, pageSize);
+        setResponses(5, 5, 0);
+        List<Result> pageOne = paginator.next();
+        assertThat(pageOne, hasSize(5));
+        assertThat(paginator.getIndexOffset(), equalTo(15));
+        List<Result> pageTwo = paginator.next();
+        assertThat(pageTwo, hasSize(5));
+        assertThat(paginator.hasNext(), equalTo(false));
+        assertThat(paginator.getBufferSize(), equalTo(0));
+        assertThat(paginator.getIndexOffset(), equalTo(20));
+
+        try {
+            paginator.next();
+        } catch (NoSuchElementException e) {
+            //do nothing
+        }
+
+        assertThat(paginator.hasNext(), equalTo(false));
+        assertThat(paginator.getBufferSize(), equalTo(0));
+        assertThat(paginator.getIndexOffset(), equalTo(20));
+    }
+
+    private void setStartIndexAndPageSize(int startIndex, int pageSize) {
+        when(queryMock.getStartIndex()).thenReturn(startIndex);
+        when(queryMock.getPageSize()).thenReturn(pageSize);
+        when(queryMock.getSortBy()).thenReturn(null);
+        when(queryMock.requestsTotalResultsCount()).thenReturn(true);
+        when(queryMock.getTimeoutMillis()).thenReturn(0L);
+        when(queryMock.getFilter()).thenReturn(mock(Filter.class));
+        paginator = new QueryResultPaginator(catalogFrameworkMock, queryMock);
+    }
+
+    private void setResponses(int... arguments) {
+        int index = 1;
+        List<List<Result>> mockResponses = new ArrayList<>();
+        for (int numberofResults : arguments) {
+            mockResponses.add(getResultListOfSize(numberofResults, index));
+            index += numberofResults;
+        }
+        List[] args = new List[arguments.length];
+        mockResponses.toArray(args);
+        List[] vaargs = Arrays.copyOfRange(args, 1, arguments.length);
+        when(queryResponseMock.getResults()).thenReturn(args[0], vaargs);
+    }
+
+    /**
+     * List of result objects populated with distance values to uniquely identify them
+     * for validation.
+     *
+     * @param pageSize    Size of result list to return for validation
+     * @param indexOffset Index value to increment to validate correct results
+     * @return Returns list of Result objects populated with distance values
+     */
+    private List<Result> getResultListOfSize(int pageSize, int indexOffset) {
+        List<Result> resultList = new ArrayList<>();
+        for (int i = 0; i < pageSize; i++) {
+            ResultImpl newResult = new ResultImpl();
+            newResult.setDistanceInMeters((double) i + indexOffset);
+            resultList.add(newResult);
+        }
+        return resultList;
+    }
+
+    /*
+     private QueryResultPaginator queryResultPaginator;
```

Comment:
Currently in progress with merging my unit tests in with Aaron's to ensure maximum coverage.